Add always and never matchers.

mithrandi opened the pull request:
Fixes LP#947026. Please criticize ruthlessly for style / formatting / etc. since I just cribbed a bunch from other modules.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/testing-cabal/testtools/207)
<!-- Reviewable:end -->

jml commented on testtools/matchers/_const.py:
```diff
@@ -0,0 +1,34 @@
+
+
+def always():
+    """Always match."""
```
Please add an example:

```
"""
Always match.

That is::

    self.assertThat(x, always())

Will always match and never fail, no matter what ``x`` is. Most useful when passed to other higher-order matchers (e.g. :py:class:`~testtools.matchers.MatchesListwise`).
"""

jml commented on testtools/matchers/_const.py:
```diff
@@ -0,0 +1,34 @@
+
+
+def never():
+    """Never match."""
```
Please add an example along the lines suggested for `always`

jml commented on testtools/matchers/_const.py:
```diff
@@ -0,0 +1,34 @@
+            _u('Inevitable mismatch on %r' % (value,)))
+
+
+def never():
```
In my head, I was thinking this would take an optional mismatch to return. Unfortunately, there's no nice way to parametrize it against the value being matched. OK to leave as-is for now. We can extend the interface later if we need to.

jml commented on testtools/tests/matchers/test_const.py:
```diff
@@ -0,0 +1,17 @@
+from testtools.matchers import Not, always, never
+
+
+class Test_ConstantMatches(TestCase):
```
Please add a docstring.

jml replied:
Also, no underscore please, just `TestConstantMatches`.

jml commented on testtools/tests/matchers/test_const.py:
```diff
@@ -0,0 +1,17 @@
+
+
+class Test_ConstantMatches(TestCase):
+    def test_always(self):
```
Please add a comment for the tests, as per https://jml.io/test-docstrings.

jml commented on testtools/tests/matchers/test_const.py:
```diff
@@ -0,0 +1,17 @@
+        self.assertThat(object(), always())
+
+    def test_never(self):
+        self.assertThat(object(), Not(never()))
```
The more normal way to write tests for matchers is to use the `TestMatchersInterface` mixin. I think these are OK as-is, but if you're feeling charitable it'd be much appreciated if you switch to that.

See example at https://github.com/testing-cabal/testtools/blob/master/testtools/tests/matchers/test_basic.py#L115

jml commented:
Thanks so much! I've been meaning to do this for years, and am very grateful you made time to do it.

Keeping up the ruthlessness…

* please add an update in `NEWS`
* if you aren't already in `AUTHORS` and the "Thanks" in `README.rst`, please add yourself (I really should automate that)
* please update `testtools/tests/matchers/__init__.py` to include the new tests (did you watch them fail?)
* please mention these in the `Combining matchers` section of the `for-test-authors.rst` document, probably just inlining the API documentation as demonstrated in `twisted-support.rst`

Ping me when you're done, and I'll give it another pass & then merge.

thomir commented:
Hi,

I wonder whether we need the `_Always` class and the `always` function? IIRC, all other matchers are title-cased, and most are classes (not that that distinction matters to the user). Is there a reason to break the established idiom of matchers appearing as classes in this case?

jml commented:
@thomir We've started to break that with the twistedsupport matchers. Basically, I don't want to export concrete types unless we have to. I've been meaning to write up my justification for this, but https://www.youtube.com/watch?v=3MNVP9-hglc is a good start.

thomir commented:
> We've started to change that idiom with the twistedsupport matchers. Basically, I don't want to export concrete types unless we have to.

Hmmmm, whether or not you're exporting a concrete type, it seems odd to me to break a naming convention. As an end user, if I want to import 'randommatcher', how do I know if I have to type:

from testtools.matchers import randommatcher

-or-

from testtools.matchers import RandomMatcher

?

I don't mind having a level of indirection via a function (that's a conversation for another day, at least), but breaking the titlecase naming convention seems like it's going to cause confusion and surprise.

rbtcollins commented:
Review status: 0 of 3 files reviewed at latest revision, 10 unresolved discussions.

---

*[testtools/matchers/__init__.py, line 18 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iB0MTMvzBQEVTQ1tg:-K9iB0MTMvzBQEVTQ1th:534113581) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/__init__.py#L18)):*
This is inconsistent with everything else; I'm open to the *idea* that we should migrate to exporting functions, but doing it in dribs and drabs with no way for folk to predict will lower usability.

---

*[testtools/matchers/__init__.py, line 49 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iBCCm4fdIk1FzHeD9:-K9iBCCm4fdIk1FzHeDA:-485711229) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/__init__.py#L49)):*
Same.

---

*[testtools/matchers/_const.py, line 16 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iBS7v4U_ggO2Ka7jO:-K9iBS7ww2OYAQ1-om5b:-126349391) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/_const.py#L16)):*
This is missing __str__, a requirement of matcher objects - see Matcher.

---

*[testtools/matchers/_const.py, line 29 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iBd6hpAB1hgPfRvM7:-K9iBd6hpAB1hgPfRvM8:411825560) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/_const.py#L29)):*
Same, missing __str__

---

*[testtools/tests/matchers/test_const.py, line 7 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9gXDX51Q3nRUejFsvt-r1-7:-K9iBlSoHccsAkzB_s4M:397400806) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/tests/matchers/test_const.py#L7)):*
FWIW we haven't previously done docstrings on test classes, and I generally consider them noise unless there is something unique about the domain being tested // manner of testing needed.

---

*[testtools/tests/matchers/test_const.py, line 8 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9gXDX51Q3nRUejFsvt-r1-8:-K9iC-yIaiiNNjq5mwOL:-2050193955) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/tests/matchers/test_const.py#L8)):*
Approximately none of our  test methods have docstrings. I don't oppose having them, but I would oppose requiring them. Docstrings are for API documentation, and tests are not (generally) an API.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/testing-cabal/testtools/207)*
<!-- Sent from Reviewable.io -->

rbtcollins commented:
Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions.

---

*[testtools/tests/matchers/test_const.py, line 12 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iCW9g2tSPO4-7TeYU:-K9iCW9h8IaTmK05qk3z:826476549) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/tests/matchers/test_const.py#L12)):*
One of the things missing is __str__ as already noted - we need tests for that.

I"d also like to see tests for the mismatch returned by never, at least in a broad interface based sense.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/testing-cabal/testtools/207)*
<!-- Sent from Reviewable.io -->

rbtcollins commented:
Thanks for the patch; have reviewed - i have some detailed nits, but conceptually sure, its a nice thing to have.

mithrandi commented:
Review status: 0 of 8 files reviewed at latest revision, 11 unresolved discussions.

---

*[testtools/matchers/__init__.py, line 18 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iB0MTMvzBQEVTQ1tg:-K9jsIP54DqfTPgfdyCK:929242880) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/__init__.py#L18)):*
I suppose you and jml need to fight this one out ;)

This isn't inconsistent with _everything_ else, though; `raises` is a function (although there is `Raises` too, which does something different), and more recently the deferred matchers also export functions rather than classes.

---

*[testtools/matchers/_const.py, line 16 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iBS7v4U_ggO2Ka7jO:-K9jtjZxlyFMyjxr2vAY:-1847154844) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/_const.py#L16)):*
Added `__str__`.

---

*[testtools/matchers/_const.py, line 20 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9gXDX51Q3nRUejFsvs-r1-20:-K9jukVIKTFipONafS1a:-499146009) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/_const.py#L20)):*
Done.

---

*[testtools/matchers/_const.py, line 29 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iBd6hpAB1hgPfRvM7:-K9jumXLBTIMD4bus4bK:1758359711) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/_const.py#L29)):*
Fixed.

---

*[testtools/matchers/_const.py, line 32 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9gXDX51Q3nRUejFsvs-r1-32:-K9juptsNM_hKBCU8LqO:1248995982) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/_const.py#L32)):*
FWIW, I thought about this a bit, but didn't come up with any great ideas.

---

*[testtools/matchers/_const.py, line 33 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9gXDX51Q3nRUejFsvs-r1-33:-K9juuEyZtE5wwqz1WV6:-499146009) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/_const.py#L33)):*
Done.

---

*[testtools/tests/matchers/test_const.py, line 7 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9gXDX51Q3nRUejFsvt-r1-7:-K9jy6uLGbAFGZ9lTL5z:801472821) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/tests/matchers/test_const.py#L7)):*
I added some (brief!) narrative docstrings, hopefully that's not too much noise.

---

*[testtools/tests/matchers/test_const.py, line 12 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9gXDX51Q3nRUejFsvt-r1-12:-K9jzksj7iHVvbI9UuEt:-1893665724) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/tests/matchers/test_const.py#L12)):*
I rewrote the tests to use `TestMatchersInterface` which seems much nicer; thanks for the pointer!

---

*[testtools/tests/matchers/test_const.py, line 12 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iCW9g2tSPO4-7TeYU:-K9jzpB6Bz7DNXzvJDqj:191192861) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/tests/matchers/test_const.py#L12)):*
I think this should now be covered by the new `TestMatchersInterface`-based tests.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/testing-cabal/testtools/207)*
<!-- Sent from Reviewable.io -->

jml commented:
@thomir We already go both ways on this (e.g. we have both `raises` and `PathExists`), so this isn't introducing a new problem. Perhaps it's exacerbating one. On balance, I would be OK with renaming these functions to have `CamelCase` names, since that seems to be the way we go most often in the code.

---

Review status: 0 of 8 files reviewed at latest revision, 7 unresolved discussions.

---

*[testtools/matchers/__init__.py, line 18 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iB0MTMvzBQEVTQ1tg:-K9l2sl3BTW7W3rj8lI9:-805284266) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/__init__.py#L18)):*
Right, there is both old & new precedent, so I'm OK with this change. As mentioned above, I'd be OK with renaming to `CamelCase`, but really don't think it's necessary. 

If we don't migrate incrementally, then we can either not migrate at all or do so en masse. Migrating en masse is _also_ a usability hit. Not migrating at all means writing worse Python, which makes our code base harder to maintain and is demotivating. Incremental migration is the least bad one, to my mind.

---

*[testtools/matchers/_const.py, line 29 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iBd6hpAB1hgPfRvM7:-K9l9h6wQ4agtCMO5L4O:-2137399250) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/_const.py#L29)):*
Thanks

---

*[testtools/matchers/_const.py, line 24 \[r2\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9l4cX9jkK_oqZcN0uT:-K9l4cX9jkK_oqZcN0uU:-1263777145) ([raw file](https://github.com/testing-cabal/testtools/blob/9fdffef8d251036c1cec9dacdc4b8b2bcca4c331/testtools/matchers/_const.py#L24)):*
Sorry, I may have mislead you. testtools docstrings don't have a newline after the initial `"""`. I can make this change when merging if necessary.

---

*[testtools/tests/matchers/test_const.py, line 8 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9gXDX51Q3nRUejFsvt-r1-8:-K9l42vZOYv5Zu_ppGo_:977946092) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/tests/matchers/test_const.py#L8)):*
I said comments, fwiw, which many tests do have. Also, I disagree that docstrings are purely for API documentation. I am OK with such comments being optional.

---

*[testtools/tests/matchers/test_const.py, line 12 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iCW9g2tSPO4-7TeYU:-K9lBgkg7UeSnBpnnojx:-1834867264) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/tests/matchers/test_const.py#L12)):*
Thanks.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-:-K9l55CE0Wburvr1qWE-:-966469357)*
<!-- Sent from Reviewable.io -->

mithrandi commented:
Review status: 0 of 8 files reviewed at latest revision, 7 unresolved discussions.

---

*[testtools/matchers/_const.py, line 24 \[r2\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9l4cX9jkK_oqZcN0uT:-K9lO-cVDZ7-f-XNXJ5K:1594491111) ([raw file](https://github.com/testing-cabal/testtools/blob/9fdffef8d251036c1cec9dacdc4b8b2bcca4c331/testtools/matchers/_const.py#L24)):*
No, that was just me switching between too many projects. Fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/testing-cabal/testtools/207)*
<!-- Sent from Reviewable.io -->

rbtcollins commented:
Ok, so since thomi and I ere both asking for Always and Never, and @jml is ok with a rename on balance; could you please do that, and then I think its landable.

---

Reviewed 1 of 3 files at r1, 5 of 7 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[testtools/matchers/__init__.py, line 18 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9iB0MTMvzBQEVTQ1tg:-K9mw4en1WAG-oFVAUGX:120558492) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/matchers/__init__.py#L18)):*
By en mass being a usability hit, I presume you mean the fact that there would be a flag day, and folk would need to remember 'before X, Upper, after X, lower' ? True - but thats a single lookup table entry, not scaled by the number of matchers we add over the next N years.

w.r.t raises, I think in hindsight we should have done _Raises for the class and Raises for the thing people use. In the world of Nouns :).

---

*[testtools/tests/matchers/test_const.py, line 8 \[r1\]](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-K9gXDX51Q3nRUejFsvt-r1-8:-K9n-mSTcEWbljqOrBxp:1916495149) ([raw file](https://github.com/testing-cabal/testtools/blob/11daa1fb16925d16267c4b9b59081cfc506d2e2a/testtools/tests/matchers/test_const.py#L8)):*
Sorry yeah, misread in the context of the request for a class docstring (which is also unusual in the test suite, but shrug).

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-:-K9n0GEi_ehnFwWHGKGy:781830998)*
<!-- Sent from Reviewable.io -->

mithrandi commented:
Okay, I uppercased the names.

---

Review status: 3 of 8 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/testing-cabal/testtools/207#-:-K9nA54GRJDwMTNRFq6P:-1501347489)*
<!-- Sent from Reviewable.io -->

thomir commented:
I'm now +1 to land - thanks!

jml commented:
Merged as f5a370e
